packages/core/src/render3/state.ts

```diff
@@ -119,27 +119,137 @@ export function getLView(): LView {
-  activeHostContext = host;
-  activeHostElementIndex = index;
+export function setActiveHostElement(elementIndex: number | null = null) {
+  if (activeDirectiveElementIndex !== elementIndex) {
+    activeDirectiveElementIndex = elementIndex == null ? -1 : elementIndex;
+    activeDirectiveUniqueId = MIN_DIRECTIVE_UNIQUE_ID;
+    activeDirectiveSuperClassDepth = 0;
+  }
 }
 
-export function getActiveHostContext() {
-  return activeHostContext;
+/**
+ * Returns the element index value of the current host element.
+ *
+ * Note that this is only active when `hostBinding` functions are being processed.
+ */
+export function getActiveDirectiveElementIndex() {
+  return activeDirectiveElementIndex;
+}
+
+/**
+ * Returns the current unique id value of the current directive.
+ *
+ * For example we have an element that has two directives on it:
+ * <div dir-one dir-two></div>
+ *
+ * dirOne->hostBindings() (uniqueId == 1)
+ * dirTwo->hostBindings() (uniqueId == 2)
+ *
+ * Note that this is only active when `hostBinding` functions are being processed.
+ *
+ * Note that directive unique id values are specific to an element (this means that
+ * the same unique id value could be present on another element with a completely
+ * different set of directives).
+ */
+export function getActiveDirectiveUniqueId() {
+  return activeDirectiveUniqueId;
 }
 
-export function getActiveHostElementIndex() {
-  return activeHostElementIndex;
+/**
+ * Increments the current directive unique id value.
+ *
+ * For example we have an element that has two directives on it:
+ * <div dir-one dir-two></div>
+ *
+ * dirOne->hostBindings() (index = 1)
+ * // increment
+ * dirTwo->hostBindings() (index = 2)
+ * 
+ * Depending on whether or not a previous directive had any inherited
+ * directives present, that value will be incremented in addition
+ * to the id jumping up by one.
+ *
+ * Note that this is only active when `hostBinding` functions are being processed.
+ *
+ * Note that directive unique id values are specific to an element (this means that
+ * the same unique id value could be present on another element with a completely
+ * different set of directives).
+ */
+export function incrementActiveDirectiveUniqueId() {
+  activeDirectiveUniqueId += 1 + activeDirectiveSuperClassDepth;
+
+  // because we are dealing with a new directive this
+  // means we have exited out of the inheritance chain
+  activeDirectiveSuperClassDepth = 0;
+}
+
+/**
+ * Increments the current super class (reverse inheritance) depth for a directive.
+ *
+ * For example we have two directives: Parent and Child.
+ * <div child-dir other-dir></div>
+ *
+ * // increment
+ * parentInstance->hostBindings() (depth = 1)
+ * // decrement
+ * childInstance->hostBindings() (depth = 0)
+ * otherInstance->hostBindings() (depth = 0 b/c it's a different directive)
+ *
+ * Note that this is only active when `hostBinding` functions are being processed.
+ */
+export function incrementActiveDirectiveSuperClassDepth() {
```

Comment:
Nit: It seems like we could combine this with the decrement function if we add a "delta" arg (1 or -1). Since these are always used together, it's more generated code to keep them separate.

```suggestion
export function updateActiveDirectiveSuperClassDepth(delta: number) {
```